sdk/react-native/src/picovoice.tsx

```diff
@@ -0,0 +1,177 @@
+type WakeWordCallback = (keywordIndex: number) => void;
+type InferenceCallback = (inference: object) => void;
+
+class Picovoice {
+  private _porcupine: Porcupine | null;
+  private _wakeWordCallback: WakeWordCallback;
+  private _rhino: Rhino | null;
+  private _inferenceCallback: InferenceCallback;
+
+  private _frameLength: number;
+  private _sampleRate: number;
+  private _version: string;
+  private _isWakeWordDetected = false;
+
+  public static async create(
+    keywordPath: string,
+    wakeWordCallback: WakeWordCallback,
+    contextPath: string,
+    inferenceCallback: InferenceCallback,
+    porcupineSensitivity: number = 0.5,
+    rhinoSensitivity: number = 0.5,
+    porcupineModelPath?: string,
+    rhinoModelPath?: string
+  ) {
+    let porcupine: Porcupine = await Porcupine.fromKeywordPaths(
+      [keywordPath],
+      porcupineModelPath,
+      [porcupineSensitivity]
+    );
+    let rhino: Rhino = await Rhino.create(
+      contextPath,
+      rhinoModelPath,
+      rhinoSensitivity
+    );
+
+    if (porcupine.frameLength !== rhino.frameLength) {
+      throw new Error('Porcupine and Rhino frame lengths are different.');
+    }
+
+    if (porcupine.sampleRate !== rhino.sampleRate) {
+      throw new Error('Porcupine and Rhino sample rates are different.');
+    }
+
+    return new Picovoice(porcupine, wakeWordCallback, rhino, inferenceCallback);
+  }
+
+  private constructor(
+    porcupine: Porcupine,
+    wakeWordCallback: WakeWordCallback,
+    rhino: Rhino,
+    inferenceCallback: InferenceCallback
+  ) {
+    this._porcupine = porcupine;
+    this._wakeWordCallback = wakeWordCallback;
+    this._rhino = rhino;
+    this._inferenceCallback = inferenceCallback;
+    this._frameLength = porcupine.frameLength;
+    this._sampleRate = porcupine.sampleRate;
+    this._version = '1.1.0';
+  }
+
+  /**
+   * Processes a frame of the incoming audio stream. Upon detection of wake word and completion of follow-on command
+   * inference invokes user-defined callbacks.
+   *
+   * @param frame A frame of audio samples. The number of samples per frame can be attained by calling
+   * `.frameLength`. The incoming audio needs to have a sample rate equal to `.sample_rate` and be 16-bit linearly-encoded.
+   * Picovoice operates on single-channel audio.
+   */
+  async process(frame: number[]) {
+    if (this._porcupine === null || this._rhino === null) {
+      throw new Error('Cannot process frame - resources have been released.');
+    }
+
+    if (!this._isWakeWordDetected) {
+      const keywordIndex = await this._porcupine.process(frame);
+
+      if (keywordIndex !== -1) {
+        this._isWakeWordDetected = true;
+        this._wakeWordCallback(keywordIndex);
+      }
+    } else {
+      const result = await this._rhino.process(frame);
+      if(result['isFinalized'] === true){
+        this._isWakeWordDetected = false;
+
+        // format result in native module did not maintain order
+        let formattedInference;
```

Comment:
is there a guarantee that this keeps the order? If we need order then we should put them in an array or just handle it in the demo manually

Reply:
This block basically just makes it more printer-friendly when using JSON stringify and removes isFinalized, which is used internally in PicovoiceManager. If we think this is unnecessary for the SDK, I can definitely move the formatting to the demo, but I still may want to remove isFinalized here to reduce confusion.